### PR TITLE
Support new flipbook node for reviews

### LINKED ISSUE #85: AY-6687_Houdini 20.5: Support new "Flipbook" node for reviews
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Please describe the feature you have in mind and explain what the current shortcomings are?

[Houdini 20.5 introduces a new ROP node "Flipbook"](https://www.sidefx.com/docs/houdini//nodes/out/flipbook.html) which acts like a replacement to the the OpenGL ROP but instead uses the Vulkan viewport renderer.

It would be great if we can add support for it to use that for "review" renders instead of the OpenGL ROP. The node's parameters pretty much seems like a 1:1 match of the OpenGL node.

### How would you imagine the implementation of the feature?

Support the new Flipbook ROP node natively as a review instance.

### Are there any labels you wish to add?

- [X] I have added the relevant labels to the enhancement request.

### Describe alternatives you've considered:

Client mentioned that for them it made sense to set a default in settings somewhere so that e.g. a review defaults to either the OpenGL node OR the Flipbook node. I'm not convinced that's the best approach yet - I think having a separate creator may be safest, that is _also_ generating `review` product type should be just fine. Then in settings the studio can enable/disable EITHER creator so the old remains as is, and the new becomes available. Since we may never know how much the two nodes will remain aligning.

### Additional context:

_No response_

### PR DESCRIPTION
**Pull request by MustafaJafar**
## Changelog Description
resolve #85 

This PR enhances the current review plugins to add support for the new flipbook node.
image

## Additional Info
Flipbook node behaves a lot similar to opengl node. 
one interesting change is the `Color Correction` setting which can choose  (`color space`) or (`display` and `view`).
which may change a bit the color space validation and collection to properly support the flipbook node.

## Testing notes:
1. Create review instance using `Review`  creator
2. Try different node types
3. Publish Reviews

### CODE REVIEW

File: client/ayon_houdini/plugins/create/create_review.py [hunk 1]
```diff
@@ -150,3 +151,10 @@ def get_pre_create_attr_defs(self):
                       minimum=0.0001,
                       decimals=3)
         ]
+
+
+class CreateFlipbookReview(CreateReview):
+    
+    identifier = "io.ayon.creators.houdini.review.flipbook"
+    label = "Review (Flipbook)"
+    node_type = "flipbook"
```
Comment (MustafaJafar) [inline client/ayon_houdini/plugins/create/create_review.py:160]: > Client mentioned that for them it made sense to set a default in settings somewhere so that e.g. a review defaults to either the OpenGL node OR the Flipbook node. I'm not convinced that's the best approach yet - I think having a separate creator may be safest, that is _also_ generating `review` product type should be just fine. Then in settings the studio can enable/disable EITHER creator so the old remains as is, and the new becomes available. Since we may never know how much the two nodes will remain aligning.

Quoted from the linked issue. 
Now, should we have a separate creator for the flipbook node? or using the same creator while exposing the node type?

**Comment by BigRoy** [comment]
> Should this PR adds support for publishing reviews from LOPs?
> Adding this for reference [ynput/OpenPype#4890](https://github.com/ynput/OpenPype/pull/4890)

I'd say that yes, we should fix that... **but** it doesn't need to be part of this PR. Instead I see that irrelevant to this because it's already relevant to the OpenGL ROP too. I thought we had an issue for that, but I can't find it 🤦

File: client/ayon_houdini/plugins/create/create_review.py (hunk 1 above)
Comment (BigRoy) [reply client/ayon_houdini/plugins/create/create_review.py:160]: Whatever makes most sense from development side of things I'd say. It may appear cleaner if it's one Creator to avoid clutter in the Create list?

**Comment by MustafaJafar** [comment]
> I thought we had an issue for that, but I can't find it 🤦 

We have this one https://github.com/ynput/ayon-houdini/issues/23

**Comment by BigRoy** [comment]
> > I thought we had an issue for that, but I can't find it 🤦
> 
> We have this one #23

Yup, saw that one - but unfortunately it's not about supporting cameras IN LOPS - but that is requesting to support `lopimportcamera` which is still at geometry level.

**Comment by BigRoy** [comment]
Let's see if we can get this one https://github.com/ynput/ayon-houdini/pull/148 merged first - so that this PR is slightly simplified and works with the changes made there.

**Comment by moonyuet** [comment]
I slightly tested on the PR and find out that it's not working for the elder version of Houdini. Shall we introduce the condition so that users would know it only supports for Houdini 20.5? Or we can do it with some scripts in elder version?
I find a simple script from this website for creating flipbook animation for the elder version of Houdini: https://www.sidefx.com/docs/houdini/anim/flipbook.html

**Comment by BigRoy** [comment]
> I slightly tested on the PR and find out that it's not working for the elder version of Houdini. Shall we introduce the condition so that users would know it only supports for Houdini 20.5? 

We could enforce the `enabled` state of the Creator to `False` in an older version of Houdini than 20.5?

> Or we can do it with some scripts in elder version?
> I find a simple script from this website for creating flipbook animation for the elder version of Houdini: https://www.sidefx.com/docs/houdini/anim/flipbook.html

No please. That flipbook would just be the `OpenGL review` we already have.

File: client/ayon_houdini/plugins/create/create_review.py (hunk 1 above)
Comment (MustafaJafar) [reply client/ayon_houdini/plugins/create/create_review.py]: https://github.com/ynput/ayon-houdini/pull/167/commits/5f8d3254c8b611c627e81f2865608637aa605fbb

File: server/settings/create.py [hunk 2]
```diff
@@ -9,6 +9,22 @@ class CreatorModel(BaseSettingsModel):
...
+    ]
+
+class CreateReviewModel(BaseSettingsModel):
+    enabled: bool = SettingsField(title="Enabled")
+    default_variants: list[str] = SettingsField(
+        title="Default Products",
+        default_factory=list,
+    )
+    node_type: str = SettingsField(
+        Title="Default Node Type",
```
Comment (MustafaJafar) [inline server/settings/create.py:25]: No idea why the title doesn't show the word `Default`
image

File: client/ayon_houdini/plugins/create/create_review.py [hunk 3]
```diff
@@ -29,8 +30,15 @@ def apply_settings(self, project_settings):
...
     def create(self, product_name, instance_data, pre_create_data):
-
-        instance_data.update({"node_type": "opengl"})
+        
+        self.node_type = pre_create_data.get("node_type")
+        hou_version = hou.applicationVersionString()
+        if self.node_type == "flipbook" and not hou_version.startswith("20.5"):
+            self.log.debug("The Flipbook node type isn't supported in Houdini versions below 20.5."
+                           " Switching to the OpenGL node type instead.")
+            self.node_type = "opengl"
```
Comment (BigRoy) [inline client/ayon_houdini/plugins/create/create_review.py:39]: Let's remove this here - and just keep the `flipbook` entry out of the `Enum`? if `houdini < 20.5`

File: client/ayon_houdini/plugins/create/create_review.py [hunk 4]
```diff
@@ -123,8 +131,15 @@ def get_pre_create_attr_defs(self):
             "bmp", "cin", "exr", "jpg", "pic", "pic.gz", "png",
             "rad", "rat", "rta", "sgi", "tga", "tif",
         ]
+        node_type_enum = [
+            "opengl", "flipbook"
+        ]
```
Comment (BigRoy) [inline client/ayon_houdini/plugins/create/create_review.py:136]:
```suggestion
        node_type_enum = ["opengl"]
        if hou.applicationVersion() >= (20, 5, 0):
            node_type_enum.append("flipbook")
```

File: client/ayon_houdini/plugins/create/create_review.py (hunk 3 above)
Comment (BigRoy) [reply client/ayon_houdini/plugins/create/create_review.py:39]: Also, we should not be updating the `self.node_type` value - it makes no sense to change the class attribute from here.

File: client/ayon_houdini/plugins/publish/extract_rop.py [hunk 5]
```diff
@@ -78,13 +78,32 @@ def update_representation_data(self,
...
 
-class ExtractOpenGL(ExtractROP,
-                    publish.ColormanagedPyblishPluginMixin):
+class ExtractOpenGLAndFlipbook(ExtractROP,
+                               publish.ColormanagedPyblishPluginMixin):
 
     order = pyblish.api.ExtractorOrder - 0.01
-    label = "Extract OpenGL"
+    label = "Extract Review (OpenGL & Flipbook)"
     families = ["rop.opengl"]
```
Comment (BigRoy) [inline client/ayon_houdini/plugins/publish/extract_rop.py:86]: I'm not too sure about both these node types going through `rop.opengl` family - they should each get their own family so that we could potentially still target just one of the two if we get to any specific validations down the line. Not critical for this PR - but worth a low prio backlog issue.

File: client/ayon_houdini/plugins/publish/extract_rop.py [hunk 6]
```diff
@@ -78,13 +78,32 @@ def update_representation_data(self,
...
+            self.log.debug("Skipping instance without instance node.")
+            return
+
+        rop_node = hou.node(instance_node)
+        if rop_node.type().name() not in {"opengl", "flipbook"}:
+            self.log.debug("Skipping OpenGl extraction. Rop node {} "
+                           "is not an OpenGl node.".format(rop_node.path()))
+            return
+
+        super(ExtractOpenGLAndFlipbook, self).process(instance)
```
Comment (BigRoy) [inline client/ayon_houdini/plugins/publish/extract_rop.py:105]:
```suggestion
        super().process(instance)
```

File: client/ayon_houdini/plugins/publish/validate_review_colorspace.py [hunk 7]
```diff
@@ -57,6 +57,18 @@ def apply_settings(cls, project_settings):
...
+
+        # This plugin is triggered when marking render as reviewable.
+        # Therefore, this plugin will run on over wrong instances.
+        # TODO: Don't run this plugin on wrong instances.
+        # This plugin should run only on review product type
+        # with instance node of opengl type.
+        if rop_node.type().name() not in {"opengl", "flipbook"}:
+            self.log.debug("Skipping Validation. Rop node {} "
+                           "is not an `opengl` or `flipbook` node.".format(rop_node.path()))
+            return
```
Comment (BigRoy) [inline client/ayon_houdini/plugins/publish/validate_review_colorspace.py:70]: Why is this needed? This is not the case anymore right? Now it triggers on `rop.opengl` family?

File: server/settings/create.py (hunk 2 above)
Comment (BigRoy) [reply server/settings/create.py:25]: It's the capital `T` ;)
```suggestion
        title="Default Node Type",
```

File: client/ayon_houdini/plugins/create/create_review.py (hunk 3 above)
Comment (MustafaJafar) [reply client/ayon_houdini/plugins/create/create_review.py]: https://github.com/ynput/ayon-houdini/pull/167/commits/9bb792ead39c988edc78f19342dcfeba8c62fa94

File: client/ayon_houdini/plugins/publish/validate_review_colorspace.py (hunk 7 above)
Comment (MustafaJafar) [reply client/ayon_houdini/plugins/publish/validate_review_colorspace.py]: https://github.com/ynput/ayon-houdini/pull/167/commits/39637766f65e6723d2a1179b4906ab0ce908f424

File: client/ayon_houdini/plugins/publish/extract_rop.py (hunk 6 above)
Comment (MustafaJafar) [reply client/ayon_houdini/plugins/publish/extract_rop.py]: https://github.com/ynput/ayon-houdini/pull/167/commits/39637766f65e6723d2a1179b4906ab0ce908f424

File: client/ayon_houdini/plugins/publish/extract_rop.py (hunk 5 above)
Comment (MustafaJafar) [reply client/ayon_houdini/plugins/publish/extract_rop.py:86]: Added a todo here https://github.com/ynput/ayon-houdini/pull/167/commits/979452c4dee0a1769be2ed418b1897467ed10524
and also here's an issue https://github.com/ynput/ayon-houdini/issues/176

File: client/ayon_houdini/api/lib.py [hunk 8]
```diff
@@ -896,40 +896,51 @@ def get_current_context_template_data_with_entity_attrs():
...
-    opengl_node.setParms(
+    node.setParms(
         {"ociocolorspace": review_color_space}
     )
 
     log.debug(
         "'OCIO Colorspace' parm on '{}' has been set to "
         "the view color space '{}'"
-        .format(opengl_node, review_color_space)
+        .format(node.path(), review_color_space)
```
Comment (moonyuet) [inline client/ayon_houdini/api/lib.py:943]: just wondering why you don't use `node` (hou.Node) here as the previous code used `opengl_node` which happened to be hou.Node.

**Comment by moonyuet** [comment]
When testing in 19.5, the flipbook option does not show up as expected.
image

File: client/ayon_houdini/api/lib.py (hunk 8 above)
Comment (MustafaJafar) [reply client/ayon_houdini/api/lib.py:943]: I thought it was already solved.

Comment (MustafaJafar) [reply client/ayon_houdini/api/lib.py:943]: for consistency, I turned this one into `node.path()` instead of just `node`
other log messages are using `node.path()`

**Review by moonyuet — CHANGES_REQUESTED** [review]
I tested with 20.5 and find out that the repair action is not repairing the colorspace param as expected and the validator keeps blocking me to publish.
The video would show you why and it is related to the regex form of the ocio colorspace in houdini i believe. i.e. `'<USE_DISPLAY_NAME>'`

https://github.com/user-attachments/assets/ec209ca0-df58-40bd-b793-1c3b064666ec

**Comment by MustafaJafar** [comment]
> The video would show you why and it is related to the regex form of the ocio colorspace in houdini i believe. i.e. `'<USE_DISPLAY_NAME>'`

I remember Roy mentioned it somewhere! 
Unfortunately, I don't know how to replicate it. I can only find the token [here](https://github.com/ynput/ayon-core/blob/a181fc897d16db9563fca73473eddee590cdd427/client/ayon_core/pipeline/colorspace.py#L1486) in the code.

**Review by moonyuet — APPROVED** [review]
Publish successfully in Houdini 20.5
image

**Comment by MustafaJafar** [comment]
> I re-test again and it seems that I haven't turned on the color configuration in Ayon setting, meaning that It only happens when the OCIO color management is not enabled and using the default Houdini ocio configuration. @BigRoy, just want to hear your thought on this. We can create an issue to resolve this later(I dont think it should be solved here but instead the separate PR.)


I was able to reproduce it in develop branch with opengl node. 
image

Actually, I can pass the validation when choosing something manually from the drop down menu. (even if I disable the color management toggle in Houdini addon settings)
image

**Comment by moonyuet** [comment]
> > I re-test again and it seems that I haven't turned on the color configuration in Ayon setting, meaning that It only happens when the OCIO color management is not enabled and using the default Houdini ocio configuration. @BigRoy, just want to hear your thought on this. We can create an issue to resolve this later(I dont think it should be solved here but instead the separate PR.)
> 
> I was able to reproduce it in develop branch with opengl node. image
> 
> Actually, I can pass the validation when choosing something manually from the drop down menu. (even if I disable the color management toggle in Houdini addon settings) image

But the users may expect that the repair action would fix it. We can actually inform the users somewhere in the docs to talk about this limitation, and we need to figure out what to do to fix it eventually if there is somebody who runs on this issue.

**Comment by MustafaJafar** [comment]
I've created an issue for it.
https://github.com/ynput/ayon-houdini/issues/179
